FilePool/FilePool/FileAccessor.h

```diff
@@ -0,0 +1,156 @@
+ * @tparam TFD
+ *  File descriptor type
+ * @details
+ *  This is a base class that hides away if the accessor is read-only or write-only,
+ *  as the locking mechanisms in each case should not bother the calling code
+ */
+template <typename TFD>
+class FileAccessor : public FileAccessorBase {
+public:
+  using ReleaseDescriptorCallback = std::function<void(TFD&&)>;
+
+  /// The wrapped file descriptor
+  TFD m_fd;
+
+  /// Destructor
+  virtual ~FileAccessor() = default;
+
+protected:
+  FileAccessor(TFD&& fd, ReleaseDescriptorCallback release_callback);
+
+  ReleaseDescriptorCallback m_release_callback;
+};
+
+/**
+ * Wraps a file descriptor together with a shared lock, so multiple read accessors pointing
+ * to the same **physical file** can exist at the same time.
+ * @tparam TFD
+ *  File descriptor type
+ * @note
+ *  The file descriptor is still unique, since normally file descriptors can not be shared
+ *  between threads (shared buffers, offsets, etc.)
+ *  What is shared is the *file* itself.
+ */
+template <typename TFD>
+class FileReadAccessor : public FileAccessor<TFD> {
+public:
+  typedef FileAccessor<TFD> Base_;
+  using ReleaseDescriptorCallback = typename Base_::ReleaseDescriptorCallback;
+  using SharedLock                = typename Base_::SharedLock;
+
+  /**
+   * Constructor
+   * @param fd
+   *    File descriptor
+   * @param release_callback
+   *    Callback to be called at destruction
+   * @param lock
+   *    Shared lock
+   */
+  FileReadAccessor(TFD&& fd, ReleaseDescriptorCallback release_callback, SharedLock lock);
+
+  /// It can not be copied
+  FileReadAccessor(const FileReadAccessor&) = delete;
+
+  /// But it can be moved
+  FileReadAccessor(FileReadAccessor&&) = default;
+
+  /// Destructor
+  virtual ~FileReadAccessor();
+
+  /// @return Always true
+  bool isReadOnly() const final;
+
+private:
+  SharedLock m_shared_lock;
+};
+
+/**
+ * Wraps a file descriptor together with an exclusive lock, so there can only be one
+ * accessor (no simultaneous reads!)
+ * @tparam TFD
+ *  File descriptor type
+ */
+template <typename TFD>
+class FileWriteAccessor : public FileAccessor<TFD> {
+public:
+  typedef FileAccessor<TFD> Base_;
+  using ReleaseDescriptorCallback = typename Base_::ReleaseDescriptorCallback;
+  using SharedLock                = typename Base_::SharedLock;
+  using UniqueLock                = typename Base_::UniqueLock;
+
+  /**
+   * Constructor
+   * @param fd
+   *    File descriptor
+   * @param release_callback
+   *    Callback to be called at destruction
+   * @param lock
+   *    Unique lock to the underlying file
+   */
+  FileWriteAccessor(TFD&& fd, ReleaseDescriptorCallback release_callback, UniqueLock lock);
+
+  /// Destructor
+  virtual ~FileWriteAccessor();
+
+  /// @return Always true
+  bool isReadOnly() const final;
+
+private:
+  UniqueLock m_unique_lock;
+};
+
+}  // namespace FilePool
+}  // namespace Euclid
+
+#define FILEACCESSOR_IMPL
+#include "_impl/FileAccessor.icpp"
```

Comment:
You can't have a relative include. I mean relative to the main header file. This is forbidden. It can only be relative to the -I directories or relative to a translation unit.
so, something like: #include "FilePool/_impl/FileAccessor.icpp"

Reply:
Thanks, replaced.

Reply:
:+1: